core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala

```diff
@@ -0,0 +1,150 @@
+package org.broadinstitute.dsde.rawls.spendreporting
+
+import java.util
+import java.util.Currency
+
+import akka.http.scaladsl.model.StatusCodes
+import com.google.api.services.bigquery.model._
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.rawls.RawlsExceptionWithErrorReport
+import org.broadinstitute.dsde.rawls.dataaccess.{SamDAO, SlickDataSource}
+import org.broadinstitute.dsde.rawls.model._
+import org.broadinstitute.dsde.workbench.google.GoogleBigQueryDAO
+import org.broadinstitute.dsde.workbench.model.google.{BigQueryDatasetName, BigQueryTableName, GoogleProject}
+import org.joda.time.{DateTime, Days}
+import org.joda.time.format.ISODateTimeFormat
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.jdk.CollectionConverters._
+import scala.math.BigDecimal.RoundingMode
+
+object SpendReportingService {
+  def constructor(dataSource: SlickDataSource, bigQueryDAO: GoogleBigQueryDAO, samDAO: SamDAO, defaultTableName: String, serviceProject: GoogleProject)
+                 (userInfo: UserInfo)
+                 (implicit executionContext: ExecutionContext): SpendReportingService = {
+    new SpendReportingService(userInfo, dataSource, bigQueryDAO, samDAO, defaultTableName, serviceProject)
+  }
+}
+
+
+class SpendReportingService(userInfo: UserInfo, dataSource: SlickDataSource, bigQueryDAO: GoogleBigQueryDAO, samDAO: SamDAO, defaultTableName: String, serviceProject: GoogleProject)
+                           (implicit val executionContext: ExecutionContext) extends LazyLogging {
+
+  def requireProjectAction[T](projectName: RawlsBillingProjectName, action: SamResourceAction)(op: => Future[T]): Future[T] = {
+    samDAO.userHasAction(SamResourceTypeNames.billingProject, projectName.value, action, userInfo).flatMap {
+      case true => op
+      case false => Future.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.Forbidden, s"You cannot perform ${action.value} on project ${projectName.value}")))
+    }
+  }
+
+  def requireAlphaUser[T]()(op: => Future[T]): Future[T] = {
+    samDAO.userHasAction(SamResourceTypeNames.managedGroup, "Alpha_Spend_Report_Users", SamResourceAction("use"), userInfo).flatMap {
+      case true => op
+      case false => Future.failed(new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, "This API is not live yet.")))
+    }
+  }
+
+  def extractSpendReportingResults(rows: List[TableRow], startTime: DateTime, endTime: DateTime): SpendReportingResults = {
+    val currency = Currency.getInstance(rows.head.getF.get(2).getV.toString)
```

Comment:
What do you think about adding an assertion that all of the rows in the result set have the same currency?

Reply:
I think that'd make sense. The `spendSummary` won't be right unless it's the same currency anyways